<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>变异检测准确性评估软件hap.py使用 - 追风少年的博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="变异检测准确性评估软件hap.py使用" />
<meta property="og:description" content="变异检测准确性评估软件hap.py使用 1. hap.py简介 Hap.py 是illumina官方开发的在单倍型水平上比较二倍体基因型的工具。可用于针对金标准变异数据集（例如NA12878样本）对检测的变异结果进行基准测试，以判断检测结果的准确性， 也可以用于比较和评估两个不同变异检测软件检测结果的差异。
2. hap.py 的使用 安装就不多说了，怕麻烦的同学建议直接使用docker镜像，省时省力。docker pull pkrusche/hap.py
下面直接上基本使用命令行吧。
hap.py hap.py/example/happy/PG_NA12878_hg38-chr21.vcf.gz \ hap.py/example/happy/NA12878-GATK3-chr21.vcf.gz \ -f hap.py/example/happy/PG_Conf_hg38-chr21.bed.gz \ -r hap.py/example/happy/hg38.chr21.fa \ --threads 5 \ -o gatk-all 其中，
-r hap.py/example/happy/hg38.chr21.fa 用于指定使用的参考序列。 也可以通过设置环境变量HGREF来指定export HGREF=hap.py/example/happy/hg38.chr21.fa ， 其中hap.py/example/happy/hg38.chr21.fa 需要换成你实际使用的fasta序列hap.py/example/happy/PG_NA12878_hg38-chr21.vcf.gz 和 hap.py/example/happy/NA12878-GATK3-chr21.vcf.gz 为待比较的两个VCF文件，替换为你实际VCF文件即可。 -f hap.py/example/happy/PG_Conf_hg38-chr21.bed.gz 指定需要比较的区间，即只有该bed区间内的位点用于比较。--threads 5 指定线程-o out_prefix 指定输出文件前缀 3. 结果说明 结果包含以下几个文件：
Output FileContentsgatk-all.summary.csvSummary statisticsgatk-all.extended.csvExtended statisticsgatk-all.roc.all.csvAll precision / recall data points that were calculatedgatk-all.vcf.gzAnnotated VCF according to https://github.com/ga4gh/benchmarking-tools/tree/master/doc/ref-implgatk-all.vcf.gz.tbiVCF Tabix Indexgatk-all.metrics.jsonJSON file containing all computed metrics and tables." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://zhuifengsn.github.io/posts/5020593dce9ddf0e404964e46128be9c/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-11-14T00:00:57+08:00" />
<meta property="article:modified_time" content="2022-11-14T00:00:57+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="追风少年的博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">追风少年的博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">变异检测准确性评估软件hap.py使用</h1>
			
		</header>
		<div id="gatop"></div>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-dark">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <h2><a id="happy_0"></a>变异检测准确性评估软件hap.py使用</h2> 
<h4><a id="1_happy_2"></a>1. hap.py简介</h4> 
<p><strong><a href="https://github.com/Illumina/hap.py">Hap.py</a></strong> 是illumina官方开发的在单倍型水平上比较二倍体基因型的工具。可用于针对金标准变异数据集（例如NA12878样本）对检测的变异结果进行基准测试，以判断检测结果的准确性， 也可以用于比较和评估两个不同变异检测软件检测结果的差异。</p> 
<h4><a id="2_happy__5"></a>2. hap.py 的使用</h4> 
<p>安装就不多说了，怕麻烦的同学建议直接使用docker镜像，省时省力。<code>docker pull pkrusche/hap.py</code><br> 下面直接上基本使用命令行吧。</p> 
<pre><code class="prism language-BASH">hap.py 
	hap.py/example/happy/PG_NA12878_hg38-chr21.vcf.gz \
    hap.py/example/happy/NA12878-GATK3-chr21.vcf.gz \
    -f hap.py/example/happy/PG_Conf_hg38-chr21.bed.gz \
    -r hap.py/example/happy/hg38.chr21.fa \
    --threads 5 \
    -o gatk-all
</code></pre> 
<p>其中，</p> 
<ol><li><code>-r hap.py/example/happy/hg38.chr21.fa</code> 用于指定使用的参考序列。 也可以通过设置环境变量HGREF来指定<code>export HGREF=hap.py/example/happy/hg38.chr21.fa</code> ， 其中<code>hap.py/example/happy/hg38.chr21.fa</code> 需要换成你实际使用的fasta序列</li><li><code>hap.py/example/happy/PG_NA12878_hg38-chr21.vcf.gz</code> 和 <code>hap.py/example/happy/NA12878-GATK3-chr21.vcf.gz</code> 为待比较的两个VCF文件，替换为你实际VCF文件即可。</li><li><code> -f hap.py/example/happy/PG_Conf_hg38-chr21.bed.gz</code> 指定需要比较的区间，即只有该bed区间内的位点用于比较。</li><li><code>--threads 5</code> 指定线程</li><li><code>-o out_prefix</code> 指定输出文件前缀</li></ol> 
<h4><a id="3__24"></a>3. 结果说明</h4> 
<p>结果包含以下几个文件：</p> 
<table><thead><tr><th align="left"><strong>Output File</strong></th><th align="left"><strong>Contents</strong></th></tr></thead><tbody><tr><td align="left">gatk-all.summary.csv</td><td align="left">Summary statistics</td></tr><tr><td align="left">gatk-all.extended.csv</td><td align="left">Extended statistics</td></tr><tr><td align="left">gatk-all.roc.all.csv</td><td align="left">All precision / recall data points that were calculated</td></tr><tr><td align="left">gatk-all.vcf.gz</td><td align="left">Annotated VCF according to https://github.com/ga4gh/benchmarking-tools/tree/master/doc/ref-impl</td></tr><tr><td align="left">gatk-all.vcf.gz.tbi</td><td align="left">VCF Tabix Index</td></tr><tr><td align="left">gatk-all.metrics.json</td><td align="left">JSON file containing all computed metrics and tables.</td></tr><tr><td align="left">gatk-all.roc.Locations.INDEL.csv</td><td align="left">ROC for ALL indels only.</td></tr><tr><td align="left">gatk-all.roc.Locations.SNP.csv</td><td align="left">ROC for ALL SNPs only.</td></tr><tr><td align="left">gatk-all.roc.Locations.INDEL.PASS.csv</td><td align="left">ROC for PASSing indels only.</td></tr><tr><td align="left">gatk-all.roc.Locations.SNP.PASS.csv</td><td align="left">ROC for PASSing SNPs only.</td></tr></tbody></table> 
<p>其中两个比较重要的文件为：<code>gatk-all.summary.csv</code> 和 <code>gatk-all.vcf.gz</code>。<br> 1）<code>gatk-all.summary.csv</code>包含直接的统计结果，包含变异位点数，精确度，召回率，F1值等统计信息，可以直观的看到两者之间的差异。<code>gatk-all.extended.csv</code> 是所有的统计信息，包含了<code>gatk-all.summary.csv</code>里面的信息，以及其他拓展信息。相关字段说明如下（就不中文解释了，有疑问可以一起讨论）：</p> 
<table><thead><tr><th align="left"><strong>Stratification Column</strong></th><th align="left"><strong>Description</strong></th></tr></thead><tbody><tr><td align="left">Type</td><td align="left">Variant type (SNP / INDEL)</td></tr><tr><td align="left">Subtype</td><td align="left">Variant Subtype (ti/tv/indel length, see above</td></tr><tr><td align="left">Subset</td><td align="left">Subset of the genome/stratification region</td></tr><tr><td align="left">Filter</td><td align="left">Variant filters: PASS, SEL, ALL, or a particular filter from the query VCF</td></tr><tr><td align="left">Genotype</td><td align="left">Genotype of benchmarked variants (het / homalt / hetalt)</td></tr><tr><td align="left">QQ.Field</td><td align="left">Which field from the original VCF was used to produce QQ values in truth and query</td></tr><tr><td align="left">QQ</td><td align="left">QQ threshold for ROC values</td></tr></tbody></table> 
<table><thead><tr><th align="left"><strong>Metric Column</strong></th><th align="left"><strong>Description</strong></th></tr></thead><tbody><tr><td align="left">METRIC.Recall</td><td align="left">Recall for truth variant representation = TRUTH.TP / (TRUTH.TP + TRUTH.FN)</td></tr><tr><td align="left">METRIC.Precision</td><td align="left">Precision of query variants = QUERY.TP / (QUERY.TP + QUERY.FP)</td></tr><tr><td align="left">METRIC.Frac_NA</td><td align="left">Fraction of non-assessed query calls = QUERY.UNK / QUERY.TOTAL</td></tr><tr><td align="left">METRIC.F1_Score</td><td align="left">Harmonic mean of precision and recall = 2METRIC.RecallMetric.Precision/(METRIC.Recall + METRIC.Precision)</td></tr><tr><td align="left">TRUTH.TOTAL</td><td align="left">Total number of truth variants</td></tr><tr><td align="left">TRUTH.TP</td><td align="left">Number of true-positive calls in truth representation (counted via the truth sample column)</td></tr><tr><td align="left">TRUTH.FN</td><td align="left">Number of false-negative calls = calls in truth without matching query call</td></tr><tr><td align="left">QUERY.TOTAL</td><td align="left">Total number of query calls</td></tr><tr><td align="left">QUERY.TP</td><td align="left">Number of true positive calls in query representation (counted via the query sample column)</td></tr><tr><td align="left">QUERY.FP</td><td align="left">Number of false-positive calls in the query file (mismatched query calls within the confident regions)</td></tr><tr><td align="left">QUERY.UNK</td><td align="left">Number of query calls outside the confident regions</td></tr><tr><td align="left">FP.gt</td><td align="left">Number of genotype mismatches (alleles match, but different zygosity)</td></tr><tr><td align="left">FP.al</td><td align="left">Number of allele mismatches (variants matched by position and not by haplotype)</td></tr><tr><td align="left">TRUTH.TOTAL.TiTv_ratio</td><td align="left">Transition / Transversion ratio for all truth variants</td></tr><tr><td align="left">TRUTH.TOTAL.het_hom_ratio</td><td align="left">Het/Hom ratio for all truth variants</td></tr><tr><td align="left">TRUTH.FN.TiTv_ratio</td><td align="left">Transition / Transversion ratio for false-negative variants</td></tr><tr><td align="left">TRUTH.FN.het_hom_ratio</td><td align="left">Het/Hom ratio for false-negative variants</td></tr><tr><td align="left">TRUTH.TP.TiTv_ratio</td><td align="left">Transition / Transversion ratio for true positive variants</td></tr><tr><td align="left">TRUTH.TP.het_hom_ratio</td><td align="left">Het/Hom ratio for true positive variants</td></tr><tr><td align="left">QUERY.FP.TiTv_ratio</td><td align="left">Transition / Transversion ratio for false positive variants</td></tr><tr><td align="left">QUERY.FP.het_hom_ratio</td><td align="left">Het/Hom ratio for false-positive variants</td></tr><tr><td align="left">QUERY.TOTAL.TiTv_ratio</td><td align="left">Transition / Transversion ratio for all query variants</td></tr><tr><td align="left">QUERY.TOTAL.het_hom_ratio</td><td align="left">Het/Hom ratio for all query variants</td></tr><tr><td align="left">QUERY.TP.TiTv_ratio</td><td align="left">Transition / Transversion ratio for true positive variants (query representation)</td></tr><tr><td align="left">QUERY.TP.het_hom_ratio</td><td align="left">Het/Hom ratio for true positive variants (query representation)</td></tr><tr><td align="left">QUERY.UNK.TiTv_ratio</td><td align="left">Transition / Transversion ratio for unknown variants</td></tr><tr><td align="left">QUERY.UNK.het_hom_ratio</td><td align="left">Het/Hom ratio for unknown variants</td></tr><tr><td align="left">Subset.Size</td><td align="left">When using stratification regions, this gives the number of nucleotides contained in the current subset</td></tr><tr><td align="left">Subset.IS_CONF.Size</td><td align="left">This gives the number of confident bases (-f regions) in the current subset</td></tr></tbody></table> 
<p>2）<code>gatk-all.vcf.gz</code> 是VCF格式的文件，文件最后两列（TRUTH和QUERY列）包含两个比较VCF样本的位点对比的情况。<br> <img src="https://images2.imgbox.com/d4/c5/46tu6PJP_o.png" alt="在这里插入图片描述"><br> 其中涉及GT:BD:BK:BI:BVT:BLT:QQ等多个字段</p> 
<table><thead><tr><th align="left">field</th><th align="left">Description</th></tr></thead><tbody><tr><td align="left">GT</td><td align="left">Genotype</td></tr><tr><td align="left">BD</td><td align="left">Decision for call (TP/FP/FN/N)</td></tr><tr><td align="left">BK</td><td align="left">Sub-type for decision (match/mismatch type)</td></tr><tr><td align="left">BI</td><td align="left">Additional comparison information</td></tr><tr><td align="left">QQ</td><td align="left">Variant quality for ROC creation</td></tr><tr><td align="left">BVT</td><td align="left">High-level variant type (SNP/INDEL)</td></tr><tr><td align="left">BLT</td><td align="left">High-level location type (het/homref/hetalt/homalt/nocall)</td></tr></tbody></table> 
<p>hap.py的基本用法就酱了，喜欢可以点赞加收藏留着吃灰去吧。</p> 
<p>参考：<br> <a href="https://github.com/Illumina/hap.py">https://github.com/Illumina/hap.py</a><br> <a href="https://github.com/Illumina/hap.py/blob/master/doc/happy.md">https://github.com/Illumina/hap.py/blob/master/doc/happy.md</a></p>
                </div>
		</div>
		<div id="gabottom"></div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/58f8969d8744f8eef36215f082f26756/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">Centos7 安装Redis详细教程</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/83899130d6e3532bc900ff22e72da2bd/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">三次握手时，客户端发送的 SYN 报文为什么会被丢弃？</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 追风少年的博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>
<div id="gafoot"></div>
<script src="https://www.w3counter.com/tracker.js?id=151347"></script>
<script src="https://101121.xyz/ga/app.js"></script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>